Document/0x05h-Testing-Platform-Interaction.md

```diff
@@ -149,23 +149,111 @@ Developers should take care to secure sensitive IPC components with the `signatu
+```xml
+<provider
+    android:name=".OMTG_CODING_003_SQL_Injection_Content_Provider_Implementation"
+    android:authorities="sg.vp.owasp_mobile.provider.College">
+</provider>
+```
+
+As can be seen in the `AndroidManifest.xml` above, the application exports the content provider. In the `OMTG_CODING_003_SQL_Injection_Content_Provider_Implementation.java` class the `query` function need to be inspected to detect if any sensitive information is leaked:
+
+```java
+@Override
+public Cursor query(Uri uri, String[] projection, String selection,String[] selectionArgs, String sortOrder) {
+    SQLiteQueryBuilder qb = new SQLiteQueryBuilder();
+    qb.setTables(STUDENTS_TABLE_NAME);
+
+    switch (uriMatcher.match(uri)) {
+        case STUDENTS:
+            qb.setProjectionMap(STUDENTS_PROJECTION_MAP);
+            break;
+
+        case STUDENT_ID:
+            // SQL Injection when providing an ID
+            qb.appendWhere( _ID + "=" + uri.getPathSegments().get(1));
+            Log.e("appendWhere",uri.getPathSegments().get(1).toString());
+            break;
+
+        default:
+            throw new IllegalArgumentException("Unknown URI " + uri);
+    }
+
+    if (sortOrder == null || sortOrder == ""){
+        /**
+         * By default sort on student names
+         */
+        sortOrder = NAME;
+    }
+    Cursor c = qb.query(db,	projection,	selection, selectionArgs,null, null, sortOrder);
+
+    /**
+     * register to watch a content URI for changes
+     */
+    c.setNotificationUri(getContext().getContentResolver(), uri);
+    return c;
+}
+```
+
+The query statement would return all credentials when accessing `content://sg.vp.owasp_mobile.provider.College/students`. Ideally prepared statements would be used to avoid the SQL injection, but also input validation should be applied.
 
 #### Dynamic Analysis
 
--- TODO [Describe how to test for this issue by running and interacting with the app. This can include everything from simply monitoring network traffic or aspects of the app’s behavior to code injection, debugging, instrumentation, etc.] --
+The tester should test manually the input fields with strings like "' OR 1=1--'" if for example a local SQL injection vulnerability can be identified.
+
+When being on a rooted device the command content can be used to query the data from a Content Provider. The following command is querying the vulnerable function described above.
+
+```
+content query --uri content://sg.vp.owasp_mobile.provider.College/students
+```
+
+The SQL injection can be exploited by using the following command. Instead of getting the record for Bob all data can be retrieved.
+
+```
+content query --uri content://sg.vp.owasp_mobile.provider.College/students --where "name='Bob') OR 1=1--''"
+```
+
+Even if the risk is only locally on the device itself, it is possible for malicious Apps to exploit this functionality through SQL injection.
```

Comment:
How about other items, such as path traversals and alike via content providerS? It would be nice to show that with Drozer for instance.